README.md
ynet_module
===========

Kernel module which implements the Y-net protocol for Yitran power line communications devices


debian/changelog
ynet (0.9.0) UNRELEASED; urgency=medium

  * Added signal quality debug attribute
  * Fixed rtnl_lock in sysfs attributes to avoid potential deadlock

 -- Kenneth Ryerson <kryerson@vermeer.com>  Tue, 21 Apr 2015 15:33:48 -0600

ynet (0.8.0) UNRELEASED; urgency=medium

  * Updated stats counters
  * Added stats reset attribute

 -- Kenneth Ryerson <kryerson@vermeer.com>  Thu, 16 Apr 2015 11:20:03 -0600

ynet (0.7.0) UNRELEASED; urgency=medium

  * Added sysfs attribute to control modulation

 -- Kenneth Ryerson <kryerson@vermeer.com>  Wed, 15 Apr 2015 14:55:38 -0600

ynet (0.6.0) UNRELEASED; urgency=medium

  * Fixed 'NOHZ local_softirq_pending 08' warnings
  * Print modulation on driver init

 -- Kenneth Ryerson <kryerson@vermeer.com>  Tue, 14 Apr 2015 09:42:50 -0600

ynet (0.5.0) UNRELEASED; urgency=medium

  * Improved handling of dropped bytes in RX packets

 -- Kenneth Ryerson <kryerson@vermeer.com>  Thu, 08 Jan 2015 08:59:49 -0600

ynet (0.4.0) UNRELEASED; urgency=medium

  * Delay sending of next packet until current is sent
  * Escape attention bytes in packet payload

 -- Kenneth Ryerson <kryerson@vermeer.com>  Mon, 01 Dec 2014 15:15:15 -0600

ynet (0.3.0) UNRELEASED; urgency=medium

  * Added error messages for failed ynet responses
  * Streamlined options for ynattach

 -- Kenneth Ryerson <kryerson@vermeer.com>  Tue, 25 Nov 2014 16:37:47 -0600

ynet (0.2.0) UNRELEASED; urgency=medium

  * Added handling for response buffer
  * Fixed parsing of incoming packets

 -- Kenneth Ryerson <kryreson@vermeer.com>  Tue, 25 Nov 2014 07:51:28 -0600

ynet (0.1.0) UNRELEASED; urgency=medium

  * Initial release.

 -- Kenneth Ryerson <kryreson@vermeer.com>  Mon, 10 Nov 2014 08:05:10 -0600


plcid.c
#include <stdio.h>
#include <sys/types.h>
#include <sys/stat.h>
#include <fcntl.h>
#include <termios.h>
#include <unistd.h>
#include <errno.h>

int plc_port = 0;

struct speed_map {
	unsigned short speed;
	unsigned short value;
};

static const struct speed_map speeds[] = {
	{B0, 0},
	{B50, 50},
	{B75, 75},
	{B110, 110},
	{B134, 134},
	{B150, 150},
	{B200, 200},
	{B300, 300},
	{B600, 600},
	{B1200, 1200},
	{B1800, 1800},
	{B2400, 2400},
	{B4800, 4800},
	{B9600, 9600},
#ifdef B19200
	{B19200, 19200},
#elif defined(EXTA)
	{EXTA, 19200},
#endif
#ifdef B38400
	{B38400, 38400/256 + 0x8000U},
#elif defined(EXTB)
	{EXTB, 38400/256 + 0x8000U},
#endif
#ifdef B57600
	{B57600, 57600/256 + 0x8000U},
#endif
#ifdef B115200
	{B115200, 115200/256 + 0x8000U},
#endif
#ifdef B230400
	{B230400, 230400/256 + 0x8000U},
#endif
#ifdef B460800
	{B460800, 460800/256 + 0x8000U},
#endif
#ifdef B921600
	{B921600, 921600/256 + 0x8000U},
#endif
};

enum { NUM_SPEEDS = sizeof(speeds)/sizeof(speeds[0]) };

static unsigned tty_baud_to_value(speed_t speed)
{
	int i = 0;

	do {
		if (speed == speeds[i].speed) {
			if (speeds[i].value & 0x8000U) {
				return ((unsigned long) (speeds[i].value) & 0x7fffU) * 256;
			}
			return speeds[i].value;
		}
	} while (++i < NUM_SPEEDS);

	return 0;
}

static speed_t tty_value_to_baud(unsigned int value)
{
	int i = 0;

	do {
		if (value == tty_baud_to_value(speeds[i].speed)) {
			return speeds[i].speed;
		}
	} while (++i < NUM_SPEEDS);

	return (speed_t) - 1;
}

int plc_open_port(char *port, speed_t baud)
{   
    /* open IT900 serial port */
    plc_port = open(port, O_RDWR | O_NOCTTY);
    if(plc_port < 0)
    {
        return errno;
    }
    
    /* set serial parameters */
    struct termios tio;
    tcgetattr(plc_port, &tio);
    
    tio.c_iflag = IGNBRK;
    tio.c_oflag = 0;
    tio.c_lflag = 0;        /* raw input mode */
    tio.c_cc[VTIME] = 0;    /* no read timer */
    tio.c_cc[VMIN] = 1;     /* block for minimum 1 byte */
    
    tio.c_cflag &= ~(CSIZE | CSTOPB | PARENB | PARODD | CRTSCTS);
    tio.c_cflag |=  CS8 | CLOCAL;
    
    cfsetispeed(&tio, baud);
    cfsetospeed(&tio, baud);
    
    tcsetattr(plc_port, TCSANOW, &tio);
    
    return 0;
}

void plc_close_port()
{
    if(plc_port)
    {
        close(plc_port);
    }
    plc_port = 0;
}

int plc_nop()
{
    int i,err;
    char request[] = {0xCA, 0x02, 0x00, 0x00, 0x00, 0x02};
    char response[7];
    char expected[] = {0xCA, 0x03, 0x00, 0x01, 0x00, 0x01, 0x05};
    
    err = write(plc_port, request, 6);
    if(err < 0)
    {
        return errno;
    }
    
    err = read(plc_port, response, 7);
    if(err < 0)
    {
        return errno;
    }
    
    for(i = 0; i < 7; i++)
    {
        if(response[i] != expected[i])
        {
            return -1;
        }
    }
    
    return 0;
}

int plc_get_id(unsigned short *netid, unsigned short *id)
{
    int i,err;
    char request[] = {0xCA, 0x07, 0x00, 0x00, 0x42, 0x06, 0x18, 0x00, 0x02, 0x00, 0x69};
    char response[11];
    char expected[] = {0xCA, 0x07, 0x00, 0x01, 0x42, 0x01};
    unsigned char checksum = 0x4B;
    
    err = write(plc_port, request, 11);
    if(err < 0)
    {
        return errno;
    }
    
    err = read(plc_port, response, 11);
    if(err < 0)
    {
        return errno;
    }
    
    for(i = 0; i < 6; i++)
    {
        if(response[i] != expected[i])
        {
            return -1;
        }
    }
    
    for(i = 6; i < 10; i++)
    {
        checksum += response[i];
    }
    
    if(checksum != response[10])
    {
        return -1;
    }
    
    *netid = response[6];
    *netid += ((unsigned short)response[7]) << 8;
    *id = response[8];
    *id += ((unsigned short)response[9]) << 8;
    
    return 0; 
}

int main(int argc, char **argv)
{
    unsigned short netid, id;
    speed_t baud;
    int tty_value = -1;
    
    if(argc != 3)
    {
        fprintf(stderr,"Usage: plcid DEVICE BAUD\n");
        return 1;
    }
    
    char *port = argv[1];
    
    tty_value = atoi(argv[2]);
	if(tty_value < 0)
	{
		fprintf(stderr,"Invalid baud rate: %s\n",argv[2]);
		return 1;
	}
	baud = tty_value_to_baud(tty_value);
	if(baud < 0)
	{
		fprintf(stderr,"Invalid baud rate: %s\n",argv[2]);
		return 1;
	}
    
    if(plc_open_port(port,baud))
    {
        fprintf(stderr,"Failed to open port\n");
        plc_close_port();
        return 1;
    }
    
    if(plc_nop())
    {
        fprintf(stderr,"NOP failed\n");
        plc_close_port();
        return 1;
    }
    
    if(plc_get_id(&netid,&id))
    {
        fprintf(stderr,"Failed to get ID\n");
        plc_close_port();
        return 1;
    }
    
    printf("%d.%d\n",netid,id);
    
    plc_close_port();
    
    return 0;
}



dkms/ynet.h
/*
 * ynet.h	Define the Y-net device driver interface and constants.
 *
 * Version:	@(#)ynet.h	0.9.0
 *
 * Author:	Kenneth Ryerson
 */
#ifndef _LINUX_YNET_H
#define _LINUX_YNET_H

/* Y-net configuration. */
#define YN_MTU		1400
#define YNET_DATA_LEN	1472
#define YNET_TIMEOUT	(2*HZ)

/* Y-net protocol characters. */
#define YNET_ATTENTION	0xCA
#define YNET_ESC        0xDB

/* Y-net packet types */
#define YNET_PACKET_TYPE_REQUEST    0x00
#define YNET_PACKET_TYPE_RESPONSE   0x01
#define YNET_PACKET_TYPE_INDICATION 0x02

/* Y-net service types */
#define YNET_SERVICE_TYPE_SHIFT			5

#define YNET_SERVICE_TYPE_EMBEDDED		0
#define YNET_SERVICE_TYPE_STACK			1
#define YNET_SERVICE_TYPE_CONFIG_MON	2
#define YNET_SERVICE_TYPE_NL_DATA		3
#define YNET_SERVICE_TYPE_MANAGEMENT_SA	5

/* Y-net opcodes */
#define YNET_OPCODE_NOP                                             0x00
#define YNET_OPCODE_GET_VERSION                                     0x01
#define YNET_OPCODE_GET_FREE_MEMORY                                 0x02
#define YNET_OPCODE_WELCOME                                         0x04
#define YNET_OPCODE_READ_FROM_NVM                                   0x05
#define YNET_OPCODE_WRITE_TO_NVM                                    0x06

#define YNET_OPCODE_RESET                                           0x20
#define YNET_OPCODE_GO_ONLINE                                       0x22
#define YNET_OPCODE_GO_OFFLINE                                      0x23

#define YNET_OPCODE_SET_PREDEFINED_PARAMETERS                       0x40
#define YNET_OPCODE_SET_DEVICE_PARAMETERS                           0x41
#define YNET_OPCODE_GET_DEVICE_PARAMETERS                           0x42
#define YNET_OPCODE_SAVE_DEVICE_PARAMETERS                          0x43
#define YNET_OPCODE_REMOTE_PARAMETERS_CHANGED                       0x4C

#define YNET_OPCODE_TX_PACKET                                       0x60
#define YNET_OPCODE_GET_NC_DATABASE_SIZE                            0x65
#define YNET_OPCODE_RX_PACKET                                       0x68
#define YNET_OPCODE_GET_NODE_INFORMATION                            0x69
#define YNET_OPCODE_DELETE_NODE_INFORMATION                         0x6A

#define YNET_OPCODE_ADMISSION_APPROVAL_RESPONSE_FROM_APPLICATION    0xA4
#define YNET_OPCODE_LEAVE_NETWORK                                   0xA6
#define YNET_OPCODE_CONNECTIVITY_STATUS_WITH_RS                     0xB1
#define YNET_OPCODE_NODE_LEFT_NETWORK                               0xB3
#define YNET_OPCODE_GET_ADMISSION_APPROVAL_FROM_APPLICATION         0xB8
#define YNET_OPCODE_ADMISSION_REFUSE                                0xB9
#define YNET_OPCODE_CONNECTED_TO_NC                                 0xBA
#define YNET_OPCODE_DISCONNECTED_FROM_NC                            0xBB
#define YNET_OPCODE_NEW_CONNECTION_TO_NC                            0xBE
#define YNET_OPCODE_NETWORK_ID_ASSIGNED                             0xBF

/* Y-net data packet flags */
#define YNET_PACKET_DATA_TYPE_INTRABCAST    0x00
#define YNET_PACKET_DATA_TYPE_INTRAUCAST    0x01
#define YNET_PACKET_DATA_TYPE_INTRAUCASTSN  0x02
#define YNET_PACKET_DATA_TYPE_INTERBCAST    0x03
#define YNET_PACKET_DATA_TYPE_INTERUCAST    0x04
#define YNET_PACKET_DATA_TTL                0x08

#define YNET_PACKET_DATA_PRIORITY_NORMAL    0x00
#define YNET_PACKET_DATA_PRIORITY_HIGH      0x01
#define YNET_PACKET_DATA_PRIORITY_EMERGENCY 0x02

#define YNET_PACKET_DATA_NOACK              0x00
#define YNET_PACKET_DATA_ACK                0x01

/* Y-net modulation types */
#define YNET_PACKET_MODULATION_AUTO         0x1F
#define YNET_PACKET_MODULATION_DCSKT_0      0x00
#define YNET_PACKET_MODULATION_DCSKT_1      0x01
#define YNET_PACKET_MODULATION_DCSKT_2      0x02
#define YNET_PACKET_MODULATION_DCSKT_3      0x03
#define YNET_PACKET_MODULATION_DCSKT_4      0x04
#define YNET_PACKET_MODULATION_DCSKT_5      0x05
#define YNET_PACKET_MODULATION_DCSKT_6      0x06
#define YNET_PACKET_MODULATION_DCSKT_7      0x07
#define YNET_PACKET_MODULATION_DCSKT_8      0x08
#define YNET_PACKET_MODULATION_DCSKT_9      0x09
#define YNET_PACKET_MODULATION_DCSKT_10     0x0A
#define YNET_PACKET_MODULATION_DCSKT_11     0x0B
#define YNET_PACKET_MODULATION_DCSKT_12     0x0C
#define YNET_PACKET_MODULATION_DCSKT_TD1    0x0D
#define YNET_PACKET_MODULATION_DCSKT_TD2    0x0E
#define YNET_PACKET_MODULATION_DCSKT_TD3    0x0F
#define YNET_PACKET_MODULATION_DCSKT_TD4    0x10
#define YNET_PACKET_MODULATION_DCSKT_TD5    0x11
#define YNET_PACKET_MODULATION_DCSKT_TD6    0x12
#define YNET_PACKET_MODULATION_DCSKT_TD7    0x13
#define YNET_PACKET_MODULATION_DCSKT_TD8    0x14
#define YNET_PACKET_MODULATION_DCSKT_TD9    0x15
#define YNET_PACKET_MODULATION_DCSKT_TD10   0x16
#define YNET_PACKET_MODULATION_DCSKT_SM     0x19
#define YNET_PACKET_MODULATION_DCSKT_RM     0x1A
#define YNET_PACKET_MODULATION_DCSKT_ERM    0x1B

/* Y-net response values */
#define YNET_RESPONSE_STATUS_FAILED                 0x00
#define YNET_RESPONSE_STATUS_SUCCESS                0x01

#define YNET_RESPONSE_STATUS_RESET_NO_EEPROM        06
#define YNET_RESPONSE_STATUS_RESET_SUCCESS          07
#define YNET_RESPONSE_STATUS_RESET_FACTORY_DEFAULTS 08
#define YNET_RESPONSE_STATUS_RESET_FATAL_ERROR      32
#define YNET_RESPONSE_STATUS_RESET_AUTO_ONLINE      64
#define YNET_RESPONSE_STATUS_RESET_SAFE_MODE        66

#define YNET_RESPONSE_DATA_RESULT1_ACCEPTED          0x00
#define YNET_RESPONSE_DATA_RESULT1_NO_MEMORY         0x01
#define YNET_RESPONSE_DATA_RESULT1_FATAL_ERROR       0x02
#define YNET_RESPONSE_DATA_RESULT1_UNKNOWN_NODE      0x03

#define YNET_RESPONSE_DATA_RESULT2_SUCCESS           0x00
#define YNET_RESPONSE_DATA_RESULT2_NA                0x01
#define YNET_RESPONSE_DATA_RESULT2_NACK              0x02
#define YNET_RESPONSE_DATA_RESULT2_NO_RESOURCES      0x03
#define YNET_RESPONSE_DATA_RESULT2_BLOCKED           0x04
#define YNET_RESPONSE_DATA_RESULT2_UNKNOWN_ERROR     0x05

#define YNET_APPLICATION_PACKET_TYPE_CONNECT         0x10
#define YNET_APPLICATION_PACKET_TYPE_DISCONNECT      0x20
#define YNET_APPLICATION_PACKET_TYPE_DATA            0x40
#define YNET_APPLICATION_PACKET_TYPE_ACK             0x80

struct ynet
{
	int magic;

	/* Various fields. */
	struct tty_struct	*tty;		/* ptr to TTY structure		*/
	struct net_device	*dev;		/* easy for intr handling	*/
	spinlock_t			lock;

	/* These are pointers to the malloc()ed frame buffers. */
	unsigned char		*rbuff;		/* receiver buffer */
	int					rcount;		/* received chars counter */
	unsigned char		*rspbuff;	/* response buffer */
	int					rspcount;	/* response chars counter */
	unsigned char		*xbuff;		/* transmitter buffer */
	unsigned char		*xhead;		/* pointer to next byte to XMIT */
	int					xleft;		/* bytes left in XMIT queue */

	/* Detailed Y-net statistics. */

	int					mtu;		/* Our mtu (to spot changes!)   */
	int					buffsize;	/* Max buffers sizes            */
	
	unsigned long		rxstate;	/* Receiver state */
#define YNS_ATTN		0			/* Attention state */
#define YNS_LENL		1			/* Length LSB */
#define YNS_LENH		2			/* Length MSB */
#define YNS_TYPE		3			/* Packet type */
#define YNS_OPCODE		4			/* Packet opcode */
#define YNS_PAYLOAD		5			/* Payload */
#define YNS_CHKSUM		6			/* Checksum byte */

	unsigned long		flags;		/* Flag values/ mode etc	 */
#define YNF_INUSE		0			/* Channel in use            */
#define YNF_ERROR		1			/* Parity, etc. error        */
#define YNF_RESP		2			/* Response received flag    */
#define YNF_DATARX		3			/* Data packet received flag */
#define YNF_RST			4			/* Reset flag				 */
#define YNF_ESCAPE      5           /* Escape flag               */
#define YNF_SQ          6			/* Signal quality flag 		 */

	unsigned char		modulation;	/* transmit modulation */
	
	unsigned short		rxlength;	/* expected incoming packet length */
	unsigned short		plidx;		/* current payload receive index */
	unsigned char		rxtype;		/* incoming packet type */
	unsigned char		rxopcode;	/* incoming packet opcode */
	unsigned char		checksum;	/* incoming packet's checksum */
	unsigned char		leased;
	dev_t				line;
	pid_t				pid;
};

#define YNET_MAGIC 0x422A

#endif	/* _LINUX_YNET_H */



update_version.py
#!/usr/bin/python

import time
import re
import sys

if __name__=="__main__":
    if len(sys.argv) != 2:
        print("usage: update_version.py VERSION")
    
    name = "Kenneth Ryerson"
    email = "kryerson@vermeer.com"
    
    version = sys.argv[1]
    
    datetime = time.strftime("%a, %d %b %Y %X")
    timezone = time.strftime("-%H%M",time.gmtime(time.timezone))
    
    #update dkms/dkms.conf
    dkms_conf_file = open("dkms/dkms.conf","r")
    dkms_conf_lines = dkms_conf_file.readlines()
    dkms_conf_file.close()
    
    dkms_conf_file = open("dkms/dkms.conf","w")
    for line in dkms_conf_lines:
        dkms_conf_file.write(re.sub("PACKAGE_VERSION=[0-9\.]+","PACKAGE_VERSION={0}".format(version),line))
    dkms_conf_file.close()
    
    print("Version info written to dkms/dkms.conf")
    
    #update dkms/ynet.c
    ynet_c_file = open("dkms/ynet.c","r")
    ynet_c_lines = ynet_c_file.readlines()
    ynet_c_file.close()
    
    ynet_c_file = open("dkms/ynet.c","w")
    for line in ynet_c_lines:
        ynet_c_file.write(re.sub("#define YNET_VERSION\t\"[0-9\.]+\"","#define YNET_VERSION\t\"{0}\"".format(version),line))
    ynet_c_file.close()
    
    print("Version info written to dkms/ynet.c")
    
    #update dkms/ynet.h
    ynet_h_file = open("dkms/ynet.h","r")
    ynet_h_lines = ynet_h_file.readlines()
    ynet_h_file.close()
    
    ynet_h_file = open("dkms/ynet.h","w")
    for line in ynet_h_lines:
        ynet_h_file.write(re.sub("Version:\t@\(#\)ynet.h\t[0-9\.]+","Version:\t@(#)ynet.h\t{0}".format(version),line))
    ynet_h_file.close()
    
    print("Version info written to dkms/ynet.h")
    
    #update debian/postinst
    postinst_file = open("debian/postinst","r")
    postinst_lines = postinst_file.readlines()
    postinst_file.close()
    
    postinst_file = open("debian/postinst","w")
    for line in postinst_lines:
        postinst_file.write(re.sub("MODVER=[0-9\.]+","MODVER={0}".format(version),line))
    postinst_file.close()
    
    print("Version info written to debian/postinst")
    
    #update debian/prerm
    prerm_file = open("debian/prerm","r")
    prerm_lines = prerm_file.readlines()
    prerm_file.close()
    
    prerm_file = open("debian/prerm","w")
    for line in prerm_lines:
        prerm_file.write(re.sub("MODVER=[0-9\.]+","MODVER={0}".format(version),line))
    prerm_file.close()
    
    print("Version info written to debian/prerm")
    
    #update debian/changelog
    debian_version_file = open("debian/changelog","r")
    debian_version_lines = debian_version_file.readlines()
    debian_version_file.close()
    
    debian_version_file = open("debian/changelog","w")
    debian_version_file.write("ynet ({0}) UNRELEASED; urgency=medium\n".format(version))
    debian_version_file.write("\n")
    debian_version_file.write("  * <Enter change here>\n")
    debian_version_file.write("\n")
    debian_version_file.write(" -- {0} <{1}>  {2} {3}\n".format(name,email,datetime,timezone))
    debian_version_file.write("\n")
    
    debian_version_file.writelines(debian_version_lines)
    debian_version_file.close()
    
    print("Version info written to debian/changelog")

    #update Makefile
    makefile_file = open("Makefile","r")
    makefile_lines = makefile_file.readlines()
    makefile_file.close()
    
    makefile_file = open("Makefile","w")
    for line in makefile_lines:
        makefile_file.write(re.sub("MODVER = [0-9\.]+","MODVER = {0}".format(version),line))
    makefile_file.close()
    
    print("Version info written to Makefile")
    


ynattach.c
/* vi: set sw=4 ts=4: */
/*
 * Attaches a Y-net protocol serial device to a network interface
 * Requires ynet kernel module
 *
 * Author: Kenneth Ryerson (kryerson at vermeer dot com)
 *
 * Licensed under GPLv2 or later, see file LICENSE in this source tree.
 *
 * Based on slattach
 *
 */

#include <stdlib.h>
#include <stdio.h>
#include <termios.h>
#include <unistd.h>
#include <sys/ioctl.h>
#include <sys/types.h>
#include <sys/stat.h>
#include <fcntl.h>
#include <string.h>
#include <malloc.h>
#include <signal.h>

/* must match the define in patched linux/tty.h */
#define N_YNET 25

static int handle;
static int saved_disc;
static struct termios saved_state;

static void syntax()
{
	printf("ynattach [-c PROG] [-s SPEED] DEVICE\n");
}

static void usage()
{
	syntax();
	printf("Attach network interface to Y-net serial device\n");
	printf("    -s SPD  Set line speed\n");
	printf("    -c PROG Run PROG when the line is hung up\n");
}

struct speed_map {
	unsigned short speed;
	unsigned short value;
};

static const struct speed_map speeds[] = {
	{B0, 0},
	{B50, 50},
	{B75, 75},
	{B110, 110},
	{B134, 134},
	{B150, 150},
	{B200, 200},
	{B300, 300},
	{B600, 600},
	{B1200, 1200},
	{B1800, 1800},
	{B2400, 2400},
	{B4800, 4800},
	{B9600, 9600},
#ifdef B19200
	{B19200, 19200},
#elif defined(EXTA)
	{EXTA, 19200},
#endif
#ifdef B38400
	{B38400, 38400/256 + 0x8000U},
#elif defined(EXTB)
	{EXTB, 38400/256 + 0x8000U},
#endif
#ifdef B57600
	{B57600, 57600/256 + 0x8000U},
#endif
#ifdef B115200
	{B115200, 115200/256 + 0x8000U},
#endif
#ifdef B230400
	{B230400, 230400/256 + 0x8000U},
#endif
#ifdef B460800
	{B460800, 460800/256 + 0x8000U},
#endif
#ifdef B921600
	{B921600, 921600/256 + 0x8000U},
#endif
};

enum { NUM_SPEEDS = sizeof(speeds)/sizeof(speeds[0]) };

static unsigned tty_baud_to_value(speed_t speed)
{
	int i = 0;

	do {
		if (speed == speeds[i].speed) {
			if (speeds[i].value & 0x8000U) {
				return ((unsigned long) (speeds[i].value) & 0x7fffU) * 256;
			}
			return speeds[i].value;
		}
	} while (++i < NUM_SPEEDS);

	return 0;
}

static speed_t tty_value_to_baud(unsigned int value)
{
	int i = 0;

	do {
		if (value == tty_baud_to_value(speeds[i].speed)) {
			return speeds[i].speed;
		}
	} while (++i < NUM_SPEEDS);

	return (speed_t) - 1;
}

/*
 * Save tty state and line discipline
 *
 * It is fine here to bail out on errors, since we haven modified anything yet
 */
static void save_state(void)
{
	/* Save line status */
	if (tcgetattr(handle, &saved_state) < 0)
	{
		fprintf(stderr,"Error getting state\n");
		exit(EXIT_FAILURE);
	}

	/* Save line discipline */
	if(ioctl(handle, TIOCGETD, &saved_disc) < 0)
	{
		fprintf(stderr,"Error getting line discipline\n");
		exit(EXIT_FAILURE);
	}
}

static int set_termios_state_or_warn(struct termios *state)
{
	int ret;

	ret = tcsetattr(handle, TCSANOW, state);
	if (ret < 0) {
		fprintf(stderr,"Error setting state\n");
		return 1; /* used as exitcode */
	}
	return 0;
}

/*
 * Restore state and line discipline for ALL managed ttys
 *
 * Restoring ALL managed ttys is the only way to have a single
 * hangup delay.
 *
 * Go on after errors: we want to restore as many controlled ttys
 * as possible.
 */
static void restore_state_and_exit(int exitcode)
{
	struct termios state;

	/* Restore line discipline */
	if (ioctl(handle, TIOCSETD, &saved_disc) < 0)
	{
		fprintf(stderr,"Failed to restore line discipline\n");
		exitcode = 1;
	}

	/* Hangup */
	memcpy(&state, &saved_state, sizeof(state));
	cfsetispeed(&state, B0);
	cfsetospeed(&state, B0);
	if (set_termios_state_or_warn(&state))
		exitcode = 1;
	sleep(1);

	/* Restore line status */
	if (set_termios_state_or_warn(&saved_state))
		exit(EXIT_FAILURE);
	
	close(handle);

	exit(exitcode);
}

/*
 * Set tty state, line discipline
 */
static void set_state(struct termios *state)
{
	int disc;

	/* Set line status */
	if (set_termios_state_or_warn(state))
		restore_state_and_exit(EXIT_FAILURE);
	/* Set line discipline (N_YNET always) */
	disc = N_YNET;
	if (ioctl(handle, TIOCSETD, &disc) < 0)
	{
		fprintf(stderr,"Failed to set line discipline\n");
		restore_state_and_exit(EXIT_FAILURE);
	}
}

static void sig_handler(int signo)
{
	restore_state_and_exit(EXIT_SUCCESS);
}

int main(int argc, char **argv)
{
	int i, c, opt;
	struct termios state;
	const char *baud_str;
	int baud_code = -1;   /* Line baud rate (system code) */
    pid_t pid, sid;

	enum {
		OPT_s_baud   = 1 << 0
	};

	/* Parse command line options */
	opt = 0;
	while((c = getopt(argc, argv, "s:")) != -1)
	{
		switch(c)
		{
		case 's':
			opt |= OPT_s_baud;
			baud_str = optarg;
			break;
		default:
			break;
		}
	}
	
	argv += optind;

	if(!*argv)
	{
		usage();
		exit(EXIT_FAILURE);
	}

	/* We want to know if the baud rate is valid before we start touching the ttys */
	if(opt & OPT_s_baud)
	{
		int tty_value = atoi(baud_str);
		if(tty_value < 0)
		{
			fprintf(stderr,"Invalid baud rate: %s\n",baud_str);
			exit(EXIT_FAILURE);
		}
		baud_code = tty_value_to_baud(tty_value);
		if (baud_code < 0)
		{
			fprintf(stderr,"Invalid baud rate: %s\n",baud_str);
			exit(EXIT_FAILURE);
		}
	}

	/* Trap signals in order to restore tty states upon exit */
	signal(SIGHUP,sig_handler);
	signal(SIGINT,sig_handler);
	signal(SIGQUIT,sig_handler);
	signal(SIGTERM,sig_handler);

	/* Open tty */
	handle = open(*argv, O_RDWR | O_NDELAY);
	if(handle < 0)
	{
		char *buf = malloc(strlen(*argv) + 5);
		strcpy(buf,"/dev/");
		strcat(buf,*argv);
		handle = open(buf, O_RDWR | O_NDELAY);
		free(buf);
		if(handle < 0)
		{
			fprintf(stderr,"Failed to open tty\n");
			exit(EXIT_FAILURE);
		}
	}

	/* Save current tty state */
	save_state();

	/* Configure tty */
	memcpy(&state, &saved_state, sizeof(state));
	
	memset(&state.c_cc, 0, sizeof(state.c_cc));
	state.c_cc[VMIN] = 1;
	state.c_iflag = IGNBRK | IGNPAR;
	state.c_oflag = 0;
	state.c_lflag = 0;
	state.c_cflag = CS8 | HUPCL | CREAD | CLOCAL;
	cfsetispeed(&state, cfgetispeed(&saved_state));
	cfsetospeed(&state, cfgetospeed(&saved_state));

	if(opt & OPT_s_baud)
	{
		cfsetispeed(&state, baud_code);
		cfsetospeed(&state, baud_code);
	}

	set_state(&state);

    /* daemonize */
    pid = fork();

    if(pid < 0)
    {
        return 1;
    }
    if(pid > 0)
    {
        return 0;
    }

    umask(0);

    sid = setsid();
    if(sid < 0)
    {
        return 1;
    }

	/* Watch line for hangup */
	while(ioctl(handle, TIOCMGET, &i) >= 0 && !(i & TIOCM_CAR))
	{
		sleep(15);
	}

	return EXIT_SUCCESS;
}

